feat: levaavev3 optimism

fp-crypto opened the pull request:
- feat: strip out flash loans
- fix: geist rewards model
- fix: most tests running
- chore: interstitial
- feat: gas optimizations
- fix: test
- fix: @jmonteer's suggestions
- fix: @rareweasel comments
- fix: minor
- fix: bad logic
- fix: revert conftest
- chore: begin work on levaave v3

fp-crypto committed:
chore: begin work on levaave v3

fp-crypto committed:
chore: more porting work

fp-crypto committed:
feat: compiles

fp-crypto committed:
feat: closer

fp-crypto committed:
feat: updates

fp-crypto committed:
feat: emode support

fp-crypto committed:
chore: cleanup

fp-crypto committed:
feat: flashloans

fp-crypto committed:
chore: add flashloan safety test

fp-crypto committed:
feat: yswaps

fp-crypto committed:
fix: address most of @Schlagonia's comments

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -1,52 +1,62 @@
+    address private constant usdc = 0x7F5c764cBc14f9669B88837ca1490cCa17c31607;
 
     // Supply and borrow tokens
     IAToken public aToken;
```
can this be immutable?

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -1,52 +1,62 @@
 
     // Supply and borrow tokens
     IAToken public aToken;
     IVariableDebtToken public debtToken;
```
same: immutable?

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -1,52 +1,62 @@
+        IVelodromeRouter(0xa132DAB612dB5cB9fC9Ac426A0Cc215A3423F9c9);
+
+    // Swap Router
+    address public router;
```
immutable?

jmonteer replied:
actually not used

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -652,69 +769,124 @@ contract Strategy is BaseStrategy {
-        if (is_weth) {
-            _path[1] = _token_out;
+        address _weth = address(weth);
+        address _usdc = address(usdc);
```
why usdc specifically? something special about velodrome?

fp-crypto replied:
Yes, most liquidity is in the usdc pairs.

jmonteer commented on contracts/Strategy.sol:
```diff
@@ -92,14 +103,16 @@ contract Strategy is BaseStrategy {
         minRewardToSell = 1e15;
 
-        router = SPOOKY_V2_ROUTER;
+        router = address(VELODROME_ROUTER);
```
not used?

fp-crypto replied:
good point

jmonteer reviewed (commented):
LGTM

fp-crypto committed:
fix: @Schlagonia's accounting suggestions

https://github.com/yearn/yearn-strategies/issues/341#issuecomment-1213446278
https://github.com/yearn/yearn-strategies/issues/341#issuecomment-1213446503

fp-crypto committed:
chore: minor cleanup

fp-crypto committed:
chore: solhint

fp-crypto committed:
fix: python version in ci

fp-crypto committed:
feat: remove uniswap references

fp-crypto committed:
feat: update requirements with newer brownie

fp-crypto committed:
feat: another try at CI

fp-crypto committed:
fix: cap the flashloan size at the token available

fp-crypto committed:
fix: minor stuff

fp-crypto committed:
feat: additional test for credit

fp-crypto committed:
feat: support setting LTV to 0

* feat: support setting LTV to 0
* fix: run black

fp-crypto committed:
chore: add some comments

fp-crypto committed:
chore: rename clone function

fp-crypto committed:
fix: assert in flashloan callback

https://github.com/yearn/yearn-strategies/issues/341#issuecomment-1235921005

fp-crypto committed:
fix: remove OZ address

https://github.com/yearn/yearn-strategies/issues/341#issuecomment-1235923198

fp-crypto committed:
fix: tests

fp-crypto committed:
fix: revoke spend when update rewards
